manager/db.go

```diff
@@ -60,128 +66,249 @@ func (m *Manager) lookupCatalog(environmentId, name string) (model.Catalog, erro
+				if err := tx.Create(&model.TemplateLabelModel{
+					TemplateLabel: model.TemplateLabel{
+						TemplateId: templateModel.ID,
+						Key:        k,
+						Value:      v,
 					},
-				}
-				if err := tx.Create(&categoryModel).Error; err != nil {
+				}).Error; err != nil {
 					tx.Rollback()
+					err := fmt.Errorf("CREATE LABEL: %v", err.Error())
 					return err
 				}
 			}
 
-			if err := tx.Create(&model.TemplateCategoryModel{
-				TemplateCategory: model.TemplateCategory{
-					TemplateId: templateModel.ID,
-					CategoryId: categoryModel.ID,
-				},
-			}).Error; err != nil {
-				tx.Rollback()
-				return err
+			if template.Category != "" {
+				// TODO: TemplateCategory composite key
+				template.Categories = append(template.Categories, template.Category)
 			}
-		}
 
-		for _, version := range template.Versions {
-			version.TemplateId = templateModel.ID
-			versionModel := model.VersionModel{
-				Version: version,
-			}
-			if err := tx.Create(&versionModel).Error; err != nil {
-				tx.Rollback()
-				return err
-			}
+			for _, category := range template.Categories {
+				var categoryModels []model.CategoryModel
+				tx.Where(&model.CategoryModel{
+					Category: model.Category{
+						Name: category,
+					},
+				}).Find(&categoryModels)
+
+				var categoryModel model.CategoryModel
+
+				categoryFound := false
+				for _, dbCategoryModel := range categoryModels {
+					if dbCategoryModel.Name == category {
+						categoryFound = true
+						categoryModel = dbCategoryModel
+						break
+					}
+				}
 
-			for k, v := range version.Labels {
-				if err := tx.Create(&model.VersionLabelModel{
-					VersionLabel: model.VersionLabel{
-						VersionId: versionModel.ID,
-						Key:       k,
-						Value:     v,
+				if !categoryFound {
+					categoryModel = model.CategoryModel{
+						Category: model.Category{
+							Name: category,
+						},
+					}
+					if err := tx.Create(&categoryModel).Error; err != nil {
+						tx.Rollback()
+						err := fmt.Errorf("CREATE CATEGORY: %v", err.Error())
+						return err
+					}
+				}
+
+				if err := tx.Create(&model.TemplateCategoryModel{
+					TemplateCategory: model.TemplateCategory{
+						TemplateId: templateModel.ID,
+						CategoryId: categoryModel.ID,
 					},
 				}).Error; err != nil {
 					tx.Rollback()
+					err := fmt.Errorf("CREATE TEMPLATE CATEGORY: %v", err.Error())
 					return err
 				}
 			}
+		}
 
-			for _, file := range version.Files {
-				file.VersionId = versionModel.ID
-				if err := tx.Create(&model.FileModel{
-					File: file,
-				}).Error; err != nil {
+		for _, version := range template.Versions {
+			if _, ok := dbVersionCommits[version.Commit]; !ok {
+				// add version to db
+				version.TemplateId = templateModel.ID
+				versionModel := model.VersionModel{
+					Version: version,
+				}
+				if err := tx.Create(&versionModel).Error; err != nil {
 					tx.Rollback()
 					return err
 				}
+
+				for k, v := range version.Labels {
```

Comment:
For version labels, they have version_id as a foreign key and have a cascaded delete. The cascade delete removes version labels and files.
 
Versions are deleted anytime there is a change.